### ISSUE TITLE
Reduce timeout on disks

### ISSUE THREAD

**Issue by Rudd-O** [zfsonlinux/zfs#471]
For every disk device that ZFS manages:

On Fri, 2011-11-25 at 11:12 +0100, e-t172 wrote:
On 2011-11-25 11:02, Fajar A. Nugraha wrote:
> > For zfs to work properly:
> > - a drive needs to be report quickly that it failed
> > - zfs needs to know that the drive failed, and kick it out of the pool
> > (possibly activating hot spare, if available)
> >
> > The problem is some drives (your HDD might be among those) will retry
> > for a LOOOONG time reading/writing to a bad sector before if it gives
> > up. And linux's driver will also spend lots of time retrying.
> 
> Maybe reducing /dev/block/*/device/timeout would alleviate this issue?
> 
> 

It'd be appropriate in case the device is just being... uhh... quite literally retarded.  Five seconds?  Ten seconds?  Should be enough in contrast to the minutes- or hours-long timeouts.

Donno if this will fix the "kernel retries reads" problem (which fundamentally we need to solve in the linux kernel), but this might help.

I'd advise to adjust the timeout even in the case of disk partitions being used rather than whole disks.  Remember that most of us who use zfs-on-root have no choice but to use ZFS on disk partitions because of /boot.

**Comment by behlendorf** [comment]
Closing.  The retry behavior in modern Linux kernels has improved and efforts to add improved logic to ZFS has been problematic.  Users who require lower timeouts should set the standard `/dev/block/*/device/timeout` tunable.

**Comment by dioni21** [comment]
I reached this issue while searching for a solution to a problem.

I have a disk that is probably near to fail, but still responds. No kernel error is logged, neither while a surface scan, nor during a scrub. But scrub find some errors.  Also, trying to read every file (rsync or zfs send) finds even more errors (100 times more than scrub).  Now doing a ddrescue copy I noticed that it sometimes freezes for some seconds (around 10s?), but continues without any error.

This disk was mirrored to another one that has failed, and I had to take off. Now I want to save my data.  If this is some kind of low ZFS timeout, how can I make this higher, just to make a full copy?

Currently using ZoL from git master, last compiled middle February, when most commits were for ZTS and FreeBSD source integration.
